File: src/nl/rubensten/texifyidea/inspections/latex/LatexUnicodeInspection.java
package nl.rubensten.texifyidea.inspections.latex;

import com.intellij.codeInsight.hint.HintManager;
import com.intellij.codeInspection.InspectionManager;
import com.intellij.codeInspection.LocalQuickFix;
import com.intellij.codeInspection.ProblemDescriptor;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.fileEditor.FileEditorManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiDocumentManager;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.util.PsiTreeUtil;
import nl.rubensten.texifyidea.insight.InsightGroup;
import nl.rubensten.texifyidea.inspections.TexifyInspectionBase;
import nl.rubensten.texifyidea.lang.Diacritic;
import nl.rubensten.texifyidea.lang.LatexCommand;
import nl.rubensten.texifyidea.lang.LatexMathCommand;
import nl.rubensten.texifyidea.lang.LatexRegularCommand;
import nl.rubensten.texifyidea.psi.LatexMathEnvironment;
import nl.rubensten.texifyidea.psi.LatexNormalText;
import nl.rubensten.texifyidea.util.Magic;
import nl.rubensten.texifyidea.util.PackageUtils;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.text.Normalizer;
import java.util.Collection;
import java.util.List;
import java.util.function.BiFunction;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/**
 * Checks whether Unicode is enabled in the file and flags illegal characters when they are not
 * supported.
 * <p>
 * Flags non-ASCII characters outside of math mode only when Unicode support packages are not
 * loaded. Unicode support is assumed when the packages {@code inputenc} and {@code fontenc} are
 * loaded. The inspection always flags non-ASCII characters in math mode, because Unicode math has
 * no support package in pdfLaTeX.
 * <p>
 * Quick fixes: <ul> <li>Escape the character: see {@link EscapeUnicodeFix}</li> <li>(When outside
 * math mode) Insert Unicode support packages: see {@link InsertUnicodePackageFix}</li> </ul>
 *
 * @author Sten Wessel
 */
public class LatexUnicodeInspection extends TexifyInspectionBase {

    private static final Pattern BASE_PATTERN = Pattern.compile("^\\p{ASCII}*");

    @NotNull
    @Override
    public InsightGroup getInspectionGroup() {
        return InsightGroup.LATEX;
    }

    @Nls
    @NotNull
    @Override
    public String getDisplayName() {
        return "Unsupported non-ASCII character";
    }

    @NotNull
    @Override
    public String getInspectionId() {
        return "Unicode";
    }

    @NotNull
    @Override
    public List<ProblemDescriptor> inspectFile(@NotNull PsiFile file, @NotNull InspectionManager
            manager, boolean isOntheFly) {

        boolean hasUnicode = unicodeEnabled(file);

        List<ProblemDescriptor> descriptors = descriptorList();

        Collection<LatexNormalText> texts = PsiTreeUtil.findChildrenOfType(file, LatexNormalText.class);
        for (LatexNormalText text : texts) {
            Matcher matcher = Magic.Pattern.nonAscii.matcher(text.getText());
            while (matcher.find()) {
                boolean inMathMode = PsiTreeUtil.getParentOfType(text, LatexMathEnvironment.class) != null;

                if (!inMathMode && hasUnicode) {
                    // Unicode is supported, characters are legal
                    continue;
                }

                descriptors.add(manager.createProblemDescriptor(
                        text,
                        new TextRange(matcher.start(), matcher.end()),
                        "Unsupported non-ASCII character",
                        ProblemHighlightType.ERROR,
                        isOntheFly,
                        new EscapeUnicodeFix(inMathMode),
                        inMathMode ? null : new InsertUnicodePackageFix()
                ));
            }
        }

        return descriptors;
    }

    /**
     * Checks whether Unicode support is enabled for the file.
     * <p>
     * Support is assumed when the packages {@code inputenc} and {@code fontenc} are loaded. The
     * loaded options are not checked.
     *
     * @param file
     *         The file to check support for.
     * @return Whether Unicode support is enabled.
     */
    static boolean unicodeEnabled(@NotNull PsiFile file) {
        // TODO: check if options are correct as well
        Collection<String> included = PackageUtils.getIncludedPackages(file);
        return Magic.Package.unicode.stream().allMatch(p -> included.contains(p.getName()));
    }

    /**
     * Inserts required packages for Unicode support.
     * <p>
     * This is only available for Unicode support outside math mode, since Unicode in math is not
     * available in pdfLaTeX.
     * <p>
     * This fix loads the packages {@code inputenc} with option {@code utf8} and {@code fontenc}
     * with option {@code T1} when needed to enable unicode support.
     */
    private static class InsertUnicodePackageFix implements LocalQuickFix {

        @Nls
        @NotNull
        @Override
        public String getFamilyName() {
            return "Include Unicode support packages";
        }

        @Override
        public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) {
            PsiElement element = descriptor.getPsiElement();
            PsiFile file = element.getContainingFile();
            Document document = PsiDocumentManager.getInstance(project).getDocument(file);
            Collection<String> included = PackageUtils.getIncludedPackages(file);

            if (document == null) {
                return;
            }

            Magic.Package.unicode.forEach(p -> {
                if (included.contains(p.getName())) {
                    return;
                }

                PackageUtils.insertUsepackage(
                        document, file, p.getName(),
                        String.join(",", p.getParameters())
                );
            });
        }
    }

    /**
     * Attempts to escape the non-ASCII character to avoid encoding issues.
     * <p>
     * The following attempts are made, in order, to determine a suitable replacement: <ol> <li> The
     * character is matched against the <em>display</em> attribute of either {@link
     * LatexRegularCommand} or {@link LatexMathCommand} (where appropiate). When there is a match,
     * the corresponding command is used as replacement. </li> <li> The character is decomposed to
     * separate combining marks (see also <a href="http://unicode.org/reports/tr15/">Unicode</a>).
     * An attempt is made to match the combining sequence against LaTeX character diacritical
     * commands. See {@link Diacritic} for a list of supported diacritics for both non-math and math
     * mode. When there is a match for all combining marks, the sequence of LaTeX commands is used
     * as replacement. Also, when the letters <em>i</em> or <em>j</em> are used in combination with
     * a diacritic their dotless versions are substituted. </li> </ol>
     * <p>
     * When neither of these steps is successful, the character is too exotic to replace and an
     * appropriate fail message is shown.
     */
    private static class EscapeUnicodeFix implements LocalQuickFix {

        private boolean inMathMode;

        EscapeUnicodeFix(boolean inMathMode) {
            this.inMathMode = inMathMode;
        }

        @Nls
        @NotNull
        @Override
        public String getFamilyName() {
            return "Escape Unicode character";
        }

        @Override
        public void applyFix(@NotNull Project project, @NotNull ProblemDescriptor descriptor) {
            PsiElement element = descriptor.getPsiElement();

            String c = descriptor.getTextRangeInElement().substring(element.getText());

            // Try to find in lookup for special command
            String replacement;
            LatexCommand command = inMathMode ? LatexMathCommand.findByDisplay(c) :
                    LatexRegularCommand.findByDisplay(c);

            // Replace with found command or with standard substitution
            if (command != null) {
                replacement = "\\" + command.getCommand();
            }
            else {
                replacement = findReplacement(c);
            }

            // When no replacement is found, show error message
            Document document = PsiDocumentManager.getInstance(project).getDocument(element.getContainingFile());
            if (replacement == null) {
                Editor editor = FileEditorManager.getInstance(project).getSelectedTextEditor();
                if (editor != null) {
                    HintManager.getInstance().showErrorHint(editor, "Character could not be converted");
                }
                return;
            }

            // Fill in replacement
            TextRange range = descriptor.getTextRangeInElement().shiftRight(element.getTextOffset());
            if (document != null) {
                document.replaceString(range.getStartOffset(), range.getEndOffset(), replacement);
            }

        }

        @Nullable
        private String findReplacement(@NotNull String c) {
            String n = Normalizer.normalize(c, Normalizer.Form.NFD);

            // Extract base characters
            Matcher matcher = BASE_PATTERN.matcher(n);
            matcher.find();
            String base = matcher.group();

            // Extract modifiers
            String[] mods = n.substring(matcher.end()).split("");

            // Diacritics
            BiFunction<String, Diacritic, String> reducer = (s, d) -> d.buildCommand(s);
            List<Diacritic> diacritics = IntStream.range(0, mods.length)
                    // Modifiers in reversed order
                    .mapToObj(i -> mods[mods.length - i - 1])
                    .map(m -> (Diacritic)(inMathMode ? Diacritic.Math.Companion.fromUnicode(m)
                            : Diacritic.Normal.Companion.fromUnicode(m)))
                    .collect(Collectors.toList());

            return Diacritic.Companion.buildChain(base, diacritics);
        }
    }
}

File: src/nl/hannahsten/texifyidea/inspections/latex/LatexUnicodeInspection.kt
package nl.hannahsten.texifyidea.inspections.latex

import com.intellij.codeInsight.hint.HintManager
import com.intellij.codeInspection.InspectionManager
import com.intellij.codeInspection.LocalQuickFix
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.options.ShowSettingsUtil
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiDocumentManager
import com.intellij.psi.PsiFile
import com.intellij.psi.util.PsiTreeUtil
import nl.hannahsten.texifyidea.insight.InsightGroup
import nl.hannahsten.texifyidea.inspections.TexifyInspectionBase
import nl.hannahsten.texifyidea.inspections.latex.LatexUnicodeInspection.EscapeUnicodeFix
import nl.hannahsten.texifyidea.inspections.latex.LatexUnicodeInspection.InsertUnicodePackageFix
import nl.hannahsten.texifyidea.lang.Diacritic
import nl.hannahsten.texifyidea.lang.LatexCommand
import nl.hannahsten.texifyidea.lang.LatexMathCommand
import nl.hannahsten.texifyidea.lang.LatexRegularCommand
import nl.hannahsten.texifyidea.psi.LatexMathEnvironment
import nl.hannahsten.texifyidea.psi.LatexNormalText
import nl.hannahsten.texifyidea.run.compiler.LatexCompiler
import nl.hannahsten.texifyidea.settings.TexifyProjectConfigurable
import nl.hannahsten.texifyidea.settings.TexifyProjectSettings
import nl.hannahsten.texifyidea.util.Magic
import nl.hannahsten.texifyidea.util.PackageUtils
import nl.hannahsten.texifyidea.util.insertUsepackage
import org.jetbrains.annotations.Nls
import java.text.Normalizer
import java.util.regex.Pattern

/**
 * Checks whether Unicode is enabled in the file and flags illegal characters when they are not
 * supported.
 *
 *
 * Flags non-ASCII characters outside of math mode only when Unicode support packages are not
 * loaded. Unicode support is assumed when the packages `inputenc` and `fontenc` are
 * loaded. The inspection always flags non-ASCII characters in math mode, because Unicode math has
 * no support package in pdfLaTeX.
 *
 * Quick fixes:
 * * Escape the character: see [EscapeUnicodeFix]
 * * (When outside math mode) Insert Unicode support packages: see [InsertUnicodePackageFix]
 *
 * @author Sten Wessel
 */
class LatexUnicodeInspection : TexifyInspectionBase() {

    companion object {
        private val BASE_PATTERN = Pattern.compile("^\\p{ASCII}*")

        /**
         * Checks whether Unicode support is enabled for the file.
         *
         * Support is assumed when the packages `inputenc` and `fontenc` are
         * loaded, or when a compiler is used which does not need these packages. The
         * loaded options are not checked.
         *
         * @param file The file to check support for.
         * @return Whether Unicode support is enabled.
         */
        internal fun unicodeEnabled(file: PsiFile): Boolean {
            if (TexifyProjectSettings.getInstance(file.project).compilerCompatibility == LatexCompiler.LUALATEX) {
                return true
            }

            // TODO: check if options are correct as well
            val included = PackageUtils.getIncludedPackages(file)
            return Magic.Package.unicode.stream().allMatch { p -> included.contains(p.name) }
        }
    }

    override val inspectionGroup = InsightGroup.LATEX

    @Nls
    override fun getDisplayName(): String {
        return "Unsupported non-ASCII character"
    }

    override val inspectionId = "Unicode"

    override fun inspectFile(file: PsiFile, manager: InspectionManager, isOntheFly: Boolean): List<ProblemDescriptor> {

        val hasUnicode = unicodeEnabled(file)

        val descriptors = descriptorList()

        val texts = PsiTreeUtil.findChildrenOfType(file, LatexNormalText::class.java)
        for (text in texts) {
            val matcher = Magic.Pattern.nonAscii.matcher(text.text)
            while (matcher.find()) {
                val inMathMode = PsiTreeUtil.getParentOfType(text, LatexMathEnvironment::class.java) != null

                if (!inMathMode && hasUnicode) {
                    // Unicode is supported, characters are legal
                    continue
                }

                descriptors.add(manager.createProblemDescriptor(
                        text,
                        TextRange(matcher.start(), matcher.end()),
                        "Unsupported non-ASCII character",
                        ProblemHighlightType.ERROR,
                        isOntheFly,
                        EscapeUnicodeFix(inMathMode),
                        if (inMathMode) {
                            null
                        }
                        else {
                            InsertUnicodePackageFix()
                        },
                        ChangeCompilerCompatibilityFix()
                ))
            }
        }

        return descriptors
    }

    /**
     * Inserts required packages for Unicode support.
     *
     *
     * This is only available for Unicode support outside math mode, since Unicode in math is not
     * available in pdfLaTeX.
     *
     *
     * This fix loads the packages `inputenc` with option `utf8` and `fontenc`
     * with option `T1` when needed to enable unicode support.
     */
    private class InsertUnicodePackageFix : LocalQuickFix {

        @Nls
        override fun getFamilyName(): String {
            return "Include Unicode support packages"
        }

        override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
            val file = descriptor.psiElement.containingFile

            Magic.Package.unicode.forEach { p ->
                file.insertUsepackage(p)
            }
        }
    }

    /**
     * Open the settings page so the user can change the compiler compability.
     */
    private class ChangeCompilerCompatibilityFix : LocalQuickFix {
        @Nls
        override fun getFamilyName(): String {
            return "Change compiler compatibility"
        }

        // We just want to open the settings window, but cannot do an AWT event inside a write action
        override fun startInWriteAction() = false

        override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
            ShowSettingsUtil.getInstance().showSettingsDialog(project, TexifyProjectConfigurable::class.java)
        }
    }

    /**
     * Attempts to escape the non-ASCII character to avoid encoding issues.
     *
     *
     * The following attempts are made, in order, to determine a suitable replacement:   1.  The
     * character is matched against the *display* attribute of either [ ] or [LatexMathCommand] (where appropiate). When there is a match,
     * the corresponding command is used as replacement.   1.  The character is decomposed to
     * separate combining marks (see also [Unicode](http://unicode.org/reports/tr15/)).
     * An attempt is made to match the combining sequence against LaTeX character diacritical
     * commands. See [Diacritic] for a list of supported diacritics for both non-math and math
     * mode. When there is a match for all combining marks, the sequence of LaTeX commands is used
     * as replacement. Also, when the letters *i* or *j* are used in combination with
     * a diacritic their dotless versions are substituted.
     *
     *
     * When neither of these steps is successful, the character is too exotic to replace and an
     * appropriate fail message is shown.
     */
    private class EscapeUnicodeFix internal constructor(private val inMathMode: Boolean) : LocalQuickFix {

        @Nls
        override fun getFamilyName(): String {
            return "Escape Unicode character"
        }

        override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
            val element = descriptor.psiElement

            val c = descriptor.textRangeInElement.substring(element.text)

            // Try to find in lookup for special command
            val replacement: String?
            val command: LatexCommand? = if (inMathMode) {
                LatexMathCommand.findByDisplay(c)
            }
            else {
                LatexRegularCommand.findByDisplay(c)
            }

            // Replace with found command or with standard substitution
            replacement = if (command != null) {
                "\\" + command.command
            }
            else {
                findReplacement(c)
            }

            // When no replacement is found, show error message
            val document = PsiDocumentManager.getInstance(project).getDocument(element.containingFile)
            if (replacement == null) {
                val editor = FileEditorManager.getInstance(project).selectedTextEditor
                if (editor != null) {
                    HintManager.getInstance().showErrorHint(editor, "Character could not be converted")
                }
                return
            }

            // Fill in replacement
            val range = descriptor.textRangeInElement.shiftRight(element.textOffset)
            document?.replaceString(range.startOffset, range.endOffset, replacement)
        }

        private fun findReplacement(c: String): String? {
            val n = Normalizer.normalize(c, Normalizer.Form.NFD)

            // Extract base characters
            val matcher = BASE_PATTERN.matcher(n)
            matcher.find()
            val base = matcher.group()

            // Extract modifiers
            val mods = n.substring(matcher.end()).split("".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

            val diacritics = (0 until mods.size)
                    // Modifiers in reversed order
                    .map { mods[mods.size - 1 - it] }
                    .map { if (inMathMode)
                        Diacritic.Math.fromUnicode(it) as Diacritic
                    else
                        Diacritic.Normal.fromUnicode(it) }

            return Diacritic.buildChain(base, diacritics)
        }
    }
}
